includes/Core/Authentication/Authentication.php

```diff
@@ -618,71 +625,80 @@ private function cron_refresh_profile_data( $user_id ) {
 	 * Handles receiving a temporary OAuth code.
 	 *
 	 * @since 1.0.0
+	 * @since n.e.x.t Moved connect and disconnect actions to dedicated handlers.
 	 */
 	private function handle_oauth() {
 		if ( defined( 'WP_CLI' ) && WP_CLI ) {
 			return;
 		}
 
-		$auth_client = $this->get_oauth_client();
-		$input       = $this->context->input();
-
 		// Handles Direct OAuth client request.
-		if ( $input->filter( INPUT_GET, 'oauth2callback' ) ) {
+		if ( $this->context->input()->filter( INPUT_GET, 'oauth2callback' ) ) {
 			if ( ! current_user_can( Permissions::AUTHENTICATE ) ) {
 				wp_die( esc_html__( 'You don\'t have permissions to authenticate with Site Kit.', 'google-site-kit' ), 403 );
 			}
 
-			$auth_client->authorize_user();
+			$this->get_oauth_client()->authorize_user();
 		}
+	}
 
-		if ( $input->filter( INPUT_GET, 'googlesitekit_disconnect' ) ) {
-			$nonce = $input->filter( INPUT_GET, 'nonce' );
-			if ( empty( $nonce ) || ! wp_verify_nonce( $nonce, 'disconnect' ) ) {
-				wp_die( esc_html__( 'Invalid nonce.', 'google-site-kit' ), 400 );
-			}
+	/**
+	 * Handles request to connect via oAuth.
+	 *
+	 * @since n.e.x.t
+	 */
+	private function handle_connect() {
+		$input = $this->context->input();
+		$nonce = $input->filter( INPUT_GET, 'nonce' );
+		if ( ! wp_verify_nonce( $nonce, self::ACTION_CONNECT ) ) {
+			wp_die( esc_html__( 'Invalid nonce.', 'google-site-kit' ), 400 );
+		}
 
-			if ( ! current_user_can( Permissions::AUTHENTICATE ) ) {
-				wp_die( esc_html__( 'You don\'t have permissions to authenticate with Site Kit.', 'google-site-kit' ), 403 );
-			}
+		if ( ! current_user_can( Permissions::AUTHENTICATE ) ) {
+			wp_die( esc_html__( 'You don\'t have permissions to authenticate with Site Kit.', 'google-site-kit' ), 403 );
+		}
 
-			$this->disconnect();
+		$redirect_url = $input->filter( INPUT_GET, 'redirect', FILTER_VALIDATE_URL );
+		if ( $redirect_url ) {
+			$redirect_url = esc_url_raw( wp_unslash( $redirect_url ) );
+		}
 
-			$redirect_url = $this->context->admin_url(
-				'splash',
-				array(
-					'googlesitekit_reset_session' => 1,
-				)
-			);
+		// User is trying to authenticate, but access token hasn't been set.
+		$additional_scopes = $input->filter( INPUT_GET, 'additional_scopes', FILTER_DEFAULT, FILTER_REQUIRE_ARRAY );
+		wp_safe_redirect(
+			esc_url_raw(
+				$this->get_oauth_client()->get_authentication_url( $redirect_url, $additional_scopes )
+			)
+		);
```

Comment:
The `esc_url_raw` call here is redundant because it is intended for database usage.

```suggestion
		wp_safe_redirect(
			$this->get_oauth_client()->get_authentication_url( $redirect_url, $additional_scopes )
		);
```

Reply:
It's not only for DB usage but rather anything other than for display. WP is rather inconsistent about its use with redirects, but there are instances of it used in core and it is mentioned in the docs:  https://developer.wordpress.org/reference/functions/esc_url_raw/#more-information

In this case, it may be unnecessary as the returned URL shouldn't have any encoded entities. I forget now but I believe the reason these were added in a number of places was from one of them biting us where it was needed. I'll take a closer look.

Reply:
Let's leave this as-is for now. We should probably re-evaluate this and other uses in a separate issue.